File: src/adapters/syft.ts
import type { NormalizedPackage } from '../types/index.js';

function formatLicense(license: any): string | undefined {
  if (!license) return undefined;
  if (typeof license === 'string') return license;
  if (Array.isArray(license)) {
    const formatted = license.map((l) => formatLicense(l)).filter(Boolean);
    return formatted.length > 0 ? formatted.join(', ') : undefined;
  }
  if (typeof license === 'object') {
    if (license.value) return license.value;
    if (license.spdxExpression) return license.spdxExpression;
    if (license.name) return license.name;
    if (license.license) return license.license;
    if (license.expression) return license.expression;
    const values = Object.values(license);
    const firstString = values.find((v) => typeof v === 'string' && v !== 'declared');
    if (firstString) return firstString as string;
  }
  return undefined;
}

export function extractSyftPackages(report: any): NormalizedPackage[] {
  const findings: NormalizedPackage[] = [];

  if (report?.artifacts) {
    for (const artifact of report.artifacts) {
      findings.push({
        source: 'syft',
        name: artifact.name,
        version: artifact.version || '',
        type: artifact.type || 'unknown',
        purl: artifact.purl ?? undefined,
        license: formatLicense(artifact.licenses),
      });
    }
  }

  return findings;
}

File: internal/adapter/syft.go
package adapter

import (
	"github.com/HarborGuard/harborguard-sensor/internal/types"
)

// ExtractSyftPackages extracts normalized packages from syft JSON.
func ExtractSyftPackages(report interface{}) []types.NormalizedPackage {
	var findings []types.NormalizedPackage

	data, ok := report.(map[string]interface{})
	if !ok {
		return findings
	}

	artifacts, ok := data["artifacts"].([]interface{})
	if !ok {
		return findings
	}

	for _, a := range artifacts {
		artifact, ok := a.(map[string]interface{})
		if !ok {
			continue
		}

		pkgType := getString(artifact, "type")
		if pkgType == "" {
			pkgType = "unknown"
		}

		findings = append(findings, types.NormalizedPackage{
			Source:  "syft",
			Name:    getString(artifact, "name"),
			Version: getString(artifact, "version"),
			Type:    pkgType,
			Purl:    getString(artifact, "purl"),
			License: formatLicense(artifact["licenses"]),
		})
	}

	return findings
}
